File: component-output-ts/componentResourceOutput.ts
import * as pulumi from "@pulumi/pulumi";

export class ComponentResourceOutput extends pulumi.ComponentResource {

    public readonly resources: pulumi.CustomResource[];

    constructor(type: string, name: string, args?: pulumi.Inputs, opts?: pulumi.ComponentResourceOptions) {
        super(type, name, args, opts);

        this.resources = [];
    }

    registerResources(resources: pulumi.CustomResource[]) {
        resources.map(resource => {
            this.registerResource(resource);
        })
    }

    registerResource(resource: pulumi.CustomResource) {
        this.resources.push(resource);
    }
}

File: component-output-py/component_resource_output.py
from pulumi import ComponentResource, CustomResource, ResourceOptions, Inputs
from pulumi_aws.ec2 import Vpc
from typing import Optional, List

class ComponentResourceOutput(ComponentResource):
    resources: List[CustomResource]

    def __init__(self, t: str, name: str, props: Optional[Inputs] = None, opts: Optional[ResourceOptions] = None, remote: bool = False) -> None:
        super().__init__(t, name, props, opts, remote)

        self.resources = []

    def register_resource(self, resource: CustomResource):
        self.resources.append(resource)

    def register_resources(self, resources: List[CustomResource]):
        for resource in resources:
            self.register_resource(resource)